### ISSUE TITLE
StateKeeper don't keep value

### ISSUE THREAD

**Issue by KangTaeJong98** [arkivanov/Decompose#469]
## Enviroment
- Android 13
- Don't keep activities On

### Compose
```kotlin
@Composable
public fun MemoRoute(
    modifier: Modifier = Modifier,
    context: MemoContext,
) {
    Children(
        modifier = modifier,
        stack = context.stack,
    ) {
        when (val instance = it.instance) {
            is MemoListContext -> {
                val memoListViewModel = instance.getViewModel(::MemoListViewModel)

                MemoListRoute(
                    memoListViewModel = memoListViewModel,
                    onAdd = context::add,
                )
            }

            is MemoAddContext -> {
                val memoAddViewModel = instance.getViewModel(::MemoAddViewModel)

                MemoAddRoute(
                    onNavigateUp = context::navigateUp,
                    memoAddViewModel = memoAddViewModel,
                )
            }

            is MemoDetailContext -> MemoDetailRoute(
                onNavigateUp = context::navigateUp,
            )
        }
    }
}
```

### ComponentContextExt
```kotlin
public inline fun <reified T : ViewModelInstance> ComponentContext.getViewModel(constructor: () -> T): T {
    return instanceKeeper.getOrCreate { constructor() }
}

public inline fun <reified T : ViewModelInstance> ComponentContext.getViewModel(constructor: (StateKeeper) -> T): T {
    return instanceKeeper.getOrCreate { constructor(stateKeeper) }
}

```

### ViewModel
```kotlin
internal class MemoAddViewModel(
    stateKeeper: StateKeeper,
) : ViewModelInstance() {
    private val _title = MutableStateFlow(stateKeeper.consume("title", ParcelableString::class)?.value.orEmpty())
    private val _description = MutableStateFlow("")

    init {
        stateKeeper.register("title") { ParcelableString(_title.value) }
    }
```

# Case1
### Only process kill : keep value

https://github.com/arkivanov/Decompose/assets/63176071/47220b91-f533-4835-aa90-fa5cc6c950d7

# Case2
### Configuration Change + process kill : don't keep value

https://github.com/arkivanov/Decompose/assets/63176071/ab6ffefa-fd63-4111-985f-accabac54362

**Comment by arkivanov** [comment]
Don't pass StateKeeper into a retained instance (aka ViewModel). You have to register with StateKeeper in your component and grab the state from the ViewModel.

**Comment by KangTaeJong98** [comment]
Can you give me idea like Android ViewModel + SavedStateHandle 🤔

**Comment by arkivanov** [comment]
That's not possible by design. You have to register a callback in StateKeeper in your component, and inside the callback take the state from ViewModel and return it for saving.

**Comment by KangTaeJong98** [comment]
Ok I'll find another way.
Thank you!

**Comment by arkivanov** [comment]
See an example: https://github.com/arkivanov/Decompose/blob/0445c326691558a3da43d8c0a45b0fa52d23629b/sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/counters/counter/DefaultCounterComponent.kt#L68

**Comment by arkivanov** [comment]
Actually, I think this should be documented.

**Comment by arkivanov** [comment]
Fixed via 90cbb10.
